Feature/co 270 minor fixes

floppyears opened the pull request.

floppyears committed:
Add pagination links

CO-270

floppyears committed:
Add JSONAPI format to single location GET request

CO-270

floppyears committed:
Update swagger file with JSONAPI format

CO-270

floppyears committed:
Update swagger to include open hours

CO-270

floppyears committed:
Add ElasticSearch health check

CO-270

floppyears committed:
Type, id and campus are case-insensitive params

CO-270

floppyears committed:
Add LocationDAO test

CO-270

This API relies on ES to do the searching. The DAO has a minimum amount
of logic. The resource class has some logic associated to JSONAPI,
but tests for that will be added in a future commit.

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/jsonapi/ResourceObject.groovy:
```diff
@@ -0,0 +1,8 @@
+    String id
+    String type
+    //@todo: how do we want to handle this???
+    def attributes
```
how about `Object attributes`?

floppyears replied:
Thanks for asking this question! I refreshed on def. Def is an alias of Object, but it has the benefit that allows us to use some of the dynamic typing. Since attributes can hold a variety of things depending on the API, I'm going to leave it as is.

browtayl replied:
you should be able to remove the comment in line 6 then

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -82,21 +84,109 @@ class LocationResource extends Resource {
-        ok(result).build()
+
+        ResultObject resultObject = new ResultObject()
+        resultObject.data = []
```
why not initialize `ResultObject.data` in the class definition?

floppyears replied:
data may be a list of a single object. When the user calls /locations - data holds a list and when the user calls /locations/123abc - data holds a single object.

browtayl replied:
understood.

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -82,21 +84,109 @@ class LocationResource extends Resource {
+
+        Integer pageNumber = getPageNumber()
+        Integer pageSize = getPageSize()
+        def map = ["q": q, "type": type, "campus": campus, "pageSize": pageSize, "pageNumber": pageNumber]
```
consider indenting for readability. also consider more descriptive variable name.

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -82,21 +84,109 @@ class LocationResource extends Resource {
+        Integer pageNumber = getPageNumber()
+        Integer pageSize = getPageSize()
+        def map = ["q": q, "type": type, "campus": campus, "pageSize": pageSize, "pageNumber": pageNumber]
+        int lastPage = Math.ceil(totalHits / pageSize)
```
verify that integer division behaves as intended. if `5 / 2 == 2`, then rounding will have no effect and a page will be missing.

browtayl replied:
modifying this line may require change or removal of lines 139-151

floppyears replied:
Good job on asking this question. I did verify with groovysh when I was writing the code. Math.ceil( 5 / 2) returns 3.0 . Groovy does the rounding off when the result of an integer division is stored in an integer.

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -82,21 +84,109 @@ class LocationResource extends Resource {
+        nonNullParams.remove('pageNumber')
+
+        // remove empty GET parameters
+        nonNullParams = nonNullParams.findAll { it.value } .collect { k, v -> "$k=$v" }
```
a loop would probably be more efficient, though the code would be less dense.

floppyears replied:
I'm going to leave this as is. Here's why:
- Groovy actually expands findAll, but it uses metaprogramming and handles nulls
- The current performance of this API is 7ms , which is super fast. 
- I'm torn on converting this groovy code into loops because it adds more code to maintain, and if we end up using katharsis , we won't be needing this code.

browtayl replied:
i was thinking in terms of operations, not milliseconds. functional programming implementations do have some performance costs. i agree that it is unlikely to be significant however.

browtayl commented on src/test/groovy/edu/oregonstate/mist/locations/frontend/LocationDAOTest.groovy:
```diff
@@ -0,0 +1,84 @@
+    public static final DropwizardAppRule<LocationsFrontendConfiguration> APPLICATION =
+            new DropwizardAppRule<LocationsFrontendConfiguration>(
+                    LocationsFrontEndApplication.class,
+                    new File("configuration.yaml").absolutePath)
```
did this line solve the issue with `configuration.yaml` in `src/main/resources`?

floppyears replied:
Yes

browtayl commented on swagger.yaml:
```diff
@@ -1,83 +1,224 @@
+
+  DayHours:
+    properties:
+      day:
```
[swagger editor](http://editor.swagger.io/#/) gives the error `Data does not match any schemas from 'anyOf'`

browtayl commented on swagger.yaml:
```diff
@@ -1,83 +1,224 @@
       responses:
-        200:
-          description: A short message
+        # Response code
```
does this resource return any status other than "ok" and "internal server error", such as "bad request"?

browtayl commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/db/LocationDAO.groovy:
```diff
@@ -46,7 +46,7 @@ class LocationDAO {
     String getById(String id) {
         try {
-            return "${ESFullUrl}/${id}/_source".toURL().text
+            return "${ESFullUrl}/${id?.toLowerCase()}/_source".toURL().text
```
will the url be correct if `id` is `null`?

browtayl replied:
why convert `String` to `URL` to `String`?

floppyears replied:
The Resource class checks that the parameter is not null. This class won't be called with a null id. If for some reason the id is null, the "?." null safe operator takes care of that. 

This method actually makes an http request . This is the short notation. In groovy, you can convert a string to a url object. Calling "text" makes an http request and returns the string response.

browtayl replied:
didn't know that. sorta like jquery. it might be worth adding a comment to make this clear to future maintainers.

browtayl commented on swagger.yaml:
```diff
@@ -1,83 +1,224 @@
+          name: is_open_dining
+          description: "Specify 1 for dining locations that are currently open"
+          required: false
+          type: string
```
`dining` and `is_open_dining` are defined as `string` but they expect the digit `1`. would `true` and `false` be more user-friendly?

floppyears replied:
Good catch. I forgot to update the parameters section. Some of these features will be implemented in the next version of the API.

floppyears committed:
Rename variable for clarity

CO-270

floppyears committed:
Minor fix based on code review

CO-270

* Removing invalid comment
* Add extra test to check getById(null)

floppyears committed:
Allow extension locations

CO-270

The backend locations now consumes extension data. ES contains
extension locations. Small change to allow searching for extension
locations.

floppyears committed:
Updates to swagger file

CO-270
